libraries/Microsoft.Bot.Builder/BotState.cs

```diff
@@ -48,84 +118,109 @@ public BotState(IStorage storage, string propertyName, Func<ITurnContext, string
         /// <returns>A task that represents the work queued to execute.</returns>
         /// <remarks>If successful, the task result contains the state object, read from storage.</remarks>
-        public virtual async Task<TState> ReadAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task LoadAsync(ITurnContext context, bool force = false, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var key = _keyDelegate(context);
-            var items = await _storage.ReadAsync(new[] { key }, cancellationToken).ConfigureAwait(false);
-            var state = items.Where(entry => entry.Key == key).Select(entry => entry.Value).OfType<TState>().FirstOrDefault();
-
-            if (state == null)
+            var cachedState = context.Services.Get<CachedBotState>(this.contextServiceKey);
+            var storageKey = storageKeyDelegate(context);
+            if (force || cachedState == null || cachedState.State == null)
             {
-                state = new TState();
+                var items = await storage.ReadAsync(new[] { storageKey }, cancellationToken).ConfigureAwait(false);
+                items.TryGetValue(storageKey, out object val);
+                context.Services[contextServiceKey] = new CachedBotState((IDictionary<String, object>)val ?? new Dictionary<string, object>());
             }
-
-            return state;
         }
 
         /// <summary>
-        /// Writes state to storage.
+        /// Writes the state object cached in the TurnContext if it is changed
         /// </summary>
         /// <param name="context">The context object for this turn.</param>
         /// <param name="state">The state object.</param>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects
         /// or threads to receive notice of cancellation.</param>
         /// <returns>A task that represents the work queued to execute.</returns>
-        public virtual async Task WriteAsync(ITurnContext context, TState state, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task SaveChangesAsync(ITurnContext context, bool force = false, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var changes = new Dictionary<string, object>();
-
-            if (state == null)
+            var cachedState = context.Services.Get<CachedBotState>(this.contextServiceKey);
+            if (force || cachedState.IsChanged())
             {
-                state = new TState();
+                var key = storageKeyDelegate(context);
+                var changes = new Dictionary<string, object>();
+                changes.Add(key, cachedState.State);
+                await this.storage.WriteAsync(changes);
+                context.Services[this.contextServiceKey] = new CachedBotState(cachedState.State);
```

Comment:
Is there actually value in re-setting to a new instance at this at this point? I'm not seeing it. Putting aside the allocation just for the new instance of the `CachedBotState`, the costs of calculating the hash again being so high when nothing in the cache would have changed seems to make it way too costly to consider.

Reply:
good catch.

---
In reply to: [203168348](https://github.com/Microsoft/botbuilder-dotnet/pull/760#discussion_r203168348) [](ancestors = 203168348)